### DISCUSSION TITLE
How to correctly programmatically push a path onto StackState

### DISCUSSION THREAD

**Discussion by thebarndog** [pointfreeco/swift-composable-architecture#2794] [Q&A] [answered]
I'm building a bit of a unique UI that's presented some navigational API challenges. Namely, my view has a header and footer that are used across the entire navigation flow and technically sit visually above the navigation stack so that they're not part of the cover transition. On the footer specifically, there's a "Next" button. The downside of that is now my header and footer aren't encapsulated inside the navigation stack and can't use `NavigationLink`s to push the next page onto the stack. So from there my thought was to dispatch a view action from the footer or header that gets intercepted by the feature reducer and from there, just push onto the stack using `StackAction`. 

The part I'm confused about though is how to create the `StackElementID`. You can't `Int`s since those are in tests and I saw in the `NavigationLink` extensions that it's being created like

```swift
@Dependency(\.stackElementID) var stackElementID
    self.init(value: state.map { StackState.Component(id: stackElementID(), element: $0) }) {
        _NavigationLinkStoreContent<P, L>(
            state: state, label: { label() }, fileID: fileID, line: line
        )
  }
```

So do I do the same thing, grab the stack element id and great a new component and grab the id via `component.id`?

**Comment by kamcma** [comment, marked as answer]
You cannot create a `StackElementID` yourself; it is deliberately an opaque type managed by `StackState`. You can only reference them once `StackState` has created them.

There are two ways to push elements onto `StackState`. One is `NavigationLink`s. The other is that `StackState` conforms to `RangeReplaceableCollection`, so you have access to methods on it such as `append(Self.Element)`. That second method sounds like what you want: your header and footer are living at the root view of the `NavigationStack`, so when a button in them sends an action to the root feature reducer, you handle that action and simply append to `StackState` programmatically, and that will cause SwiftUI to push the corresponding view onto the `NavigationStack`.

**Reply by thebarndog** [reply]
If I use `append` is that sufficient? I assumed that not adding an id might cause some undefined behavior

**Reply by kamcma** [reply]
Yes, `StackState` will take what you append to it and generate an ID for it internally. In production, it uses `UUID`s, and in tests it uses generational indices (1, 2, 3, 4, etc). It happens here in this method on `StackState`, which `append` uses under the hood, where it calls out to the `StackElementIDGenerator` dependency and calls `next()` on it.

https://github.com/pointfreeco/swift-composable-architecture/blob/de63b1d436a4d6fcc8c39f23fe4eb9a10b867a52/Sources/ComposableArchitecture/Reducer/Reducers/StackReducer.swift#L165-L171

The docs mention this as well, [here](https://pointfreeco.github.io/swift-composable-architecture/1.8.0/documentation/composablearchitecture/stackelementid#overview).

**Reply by thebarndog** [reply]
Yeah I saw the bit about using integers in tests but that it wouldn't work in production; was not aware of the id iterator. That's pretty cool!

I think what tripped me up is I assumed given how this entire architecture is driven by dispatching actions, that triggering a navigation state change would function in the same way where a view could just say `send(.push(.newPage))`. Functionally it boils down to the same thing if you dispatch your own action and then append, just a bit more indirection. 

Thank you for the answer! That does help quite a bit.
